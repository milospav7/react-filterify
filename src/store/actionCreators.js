import {
  OVERRIDE_STATE,
  RESET_ALL_FILTERS,
  RESET_POPERTY_FILTERS,
  UPDATE_FILTER,
  UPDATE_FUNCTION_FILTER,
  UPDATE_MULTIPLE_FUNCTION_FILTERS,
  UPDATE_NAVIGATION_PROPERTY_FILTER,
} from "./actionTypes";

/**
 * Updates propertyFilters map inside filter instance
 * @param {*} id redux filter Id - must be initialized inside gridFilters.js reducer
 * @param {*} filteringProperty name of the property/field used for filtering. If contains ';' char then it will be processed as multi filter - filtering more than one field with one common filter(value). Example: 'Width;Height'
 * @param {*} filteringValue value used for filtering - both number/string and array of number/string are supported
 * @param {*} operator operator used for field filtering - contains, eq, neq etc..
 * @param {*} logic operator used as logic operator between multiple values..
 */
export const updatePropertyFilter = (
  id,
  filteringProperty,
  filteringValue,
  operator = "eq",
  logic = "or",
  allowNullValue = false
) => ({
  type: UPDATE_FILTER,
  id,
  filteringProperty,
  filteringValue,
  operator,
  logic,
  allowNullValue,
});

/**
 * Resets all in filter instance, reverts to initial state
 * @param {*} id redux filter Id - must be initialized inside gridFilters.js reducer
 */
export const resetAllFilters = (id, keepTreeViewFilter = false) => ({
  type: RESET_ALL_FILTERS,
  id,
  keepTreeViewFilter,
});

/**
 * Resets all in filter instance, reverts to initial state
 * @param {*} id redux filter Id - must be initialized inside gridFilters.js reducer
 * @param {*} filteringProperties collections of filter names that should be reseted
 */
export const resetPropertyFiltersByNames = (id, filteringProperties) => ({
  type: RESET_POPERTY_FILTERS,
  id,
  filteringProperties,
});

/**
 *	Updates navigationPropetyFilters map inside filter instance
 * @param {*} id redux filter Id - must be initialized inside gridFilters.js reducer
 * @param {*} navigationProperty name of collection based property that we want to filter with lambda expression
 * @param {*} filteringProperty name of the property/field of navigation entity, used for filtering inside lambda expression
 * @param {*} filteringValue value used for filtering - both number/string and array of number/string are supported
 * @param {*} generatedExpression use only when you want to pass functions inside lambda expression(when filteringValue argument is not good enough) - for example if we want to include 'inCachedView(s/SubstationID, 21fdb7bb-39db-435d-af9a-60d695303ad1, true)' in navigation property lambda, then we need to pass this 'inCachedView(s/SubstationID, 21fdb7bb-39db-435d-af9a-60d695303ad1, true)' using generatedExpression argument
 * @param {*} navigationPropertyIsNested indicates whether provided navigation property name represents nav property inside nav property(important when constructing query string, because if it is nested it means that two lambdas will be created for this filter - one inside another). Example: navigationProperty: 'Users/Books', this is nested and will generate smth like '...Users/any(u: Books/any(b: andHereComesgeneratedExpressionOrfilteringValuesWithOperators))..' At the moment we support maximum of 2 level nesting.
 * @example1 updateNavigationPropertyFilter(12, 'RelatedSubstations', 'OperationalRegion', 'OpZone 1', null) - for final filter result(after using gridFilter reducer helper methods): RelatedSubstations/any(s: ( s
 * @example2 updateNavigationPropertyFilter(12, 'RelatedSubstations', 'OperationalRegion', '21fdb7bb-39db-435d-af9a-60d695303ad1', 'inCachedView(s/SubstationID, 21fdb7bb-39db-435d-af9a-60d695303ad1, true)') for final filter result(after using gridFilter reducer helper methods): RelatedSubstations/any(s: inCachedView(s/SubstationID, 21fdb7bb-39db-435d-af9a-60d695303ad , true)), /OperationalRegion eq 'OpZone 1')
 * @note There is one limitation when using generatedExpression argument. If you need lambda input parameter inside passed expression, use 's' for lambda input-paramter, example: 'isNotEmpty(s.Name)',
 */
export const updateNavigationPropertyFilter = (
  id,
  navigationProperty,
  filteringProperty,
  filteringValue,
  generatedExpression = null,
  navigationPropertyIsNested = false
) => ({
  type: UPDATE_NAVIGATION_PROPERTY_FILTER,
  id,
  navigationProperty,
  filteringProperty,
  filteringValue,
  generatedExpression,
  navigationPropertyIsNested,
});

/**
 * Updates functionFilters array inside filter instance
 * @param {*} id redux filter Id - must be initialized inside gridFilters.js reducer
 * @param {*} filteringProperty unique identifier of function filter string. Should be field name
 * @param {*} filterQueryString function filter that needs to be included in query string - non standard functions
 * @param {*} values this can be used as a container for values that are provided inside function query string, as a separete object for putting raw values only. Format/structure is not strict, and should be constrained only by the way of retreiving these values back in caller component - for example when we need to read these values on component mount we can use this object from redux(because we can't read them from raw query string)
 * @example filterQueryString: 'insideTimeFrame(PlannedStartTime, PlannedEndTime, 2020-06-30T22:00:00.000Z, null, false, false)'
 */
export const updateFunctionFilter = (
  id,
  filteringProperty,
  filterQueryString,
  values = null
) => ({
  type: UPDATE_FUNCTION_FILTER,
  id,
  filteringProperty,
  filterQueryString,
  values,
});

export const updateMultipleFunctionFilters = (id, filtersToUpdate) => ({
  type: UPDATE_MULTIPLE_FUNCTION_FILTERS,
  id,
  filtersToUpdate,
});

export const overrideFilterState = (id, filterState) => ({
  type: OVERRIDE_STATE,
  id,
  filterState,
});
