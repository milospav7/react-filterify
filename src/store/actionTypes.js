export const FILTER_UPDATE_FILTER = "[SHARED] FILTER_UPDATE_FILTER";
export const FILTER_RESET_ALL_FILTERS = "[SHARED] FILTER_RESET_ALL_FILTERS";
export const FILTER_RESET_POPERTY_FILTERS =
  "[SHARED] FILTER_RESET_POPERTY_FILTERS";
export const FILTER_UPDATE_NAVIGATION_PROPERTY_FILTER =
  "[SHARED] FILTER_UPDATE_NAVIGATION_PROPERTY_FILTER";
export const FILTER_UPDATE_FUNCTION_FILTER =
  "[SHARED] FILTER_UPDATE_FUNCTION_FILTER";
export const FILTER_UPDATE_MULTIPLE_FUNCTION_FILTERS =
  "[SHARED] FILTER_UPDATE_MULTIPLE_FUNCTION_FILTERS";
export const FILTER_OVERRIDE_FILTER_STATE =
  "[SHARED] FILTER_OVERRIDE_FILTER_STATE";
