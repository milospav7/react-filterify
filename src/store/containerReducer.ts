import { serializeFilter } from "../utils/odata-filtering.operators";
import {
  UPDATE_FILTER,
  RESET_ALL_FILTERS,
  UPDATE_NAVIGATION_PROPERTY_FILTER,
  UPDATE_FUNCTION_FILTER,
  RESET_POPERTY_FILTERS,
  OVERRIDE_STATE,
  UPDATE_MULTIPLE_FUNCTION_FILTERS,
} from "./actionTypes";
import {
  cloneDeep,
  valueShouldBeRemoved,
  isBoolean,
  removeKey,
  concatFilterQuerySubstrings,
} from "./helpers";
import { ContainerType } from "./types";

export const containerInitialState: ContainerType = {
  propertyFilters: {},
  navigationPropertyFilters: {},
  functionFilters: [],
};

export const containerReducer = (
  state = containerInitialState,
  action: any
) => {
  switch (action.type) {
    case UPDATE_FILTER: {
      const {
        filteringProperty,
        filteringValue,
        operator,
        logic,
        allowNullValue,
      } = action;
      const shouldRemoveFilter =
        !allowNullValue && valueShouldBeRemoved(filteringValue);

      if (shouldRemoveFilter) {
        return {
          ...state,
          propertyFilters: {
            ...removeKey(filteringProperty, state.propertyFilters),
          },
        };
      }

      return {
        ...state,
        propertyFilters: {
          ...state.propertyFilters,
          [filteringProperty]: {
            value: filteringValue,
            operator,
            logic,
          },
        },
      };
    }
    case UPDATE_NAVIGATION_PROPERTY_FILTER: {
      const {
        navigationProperty,
        filteringProperty,
        filteringValue,
        generatedExpression,
        navigationPropertyIsNested,
      } = action;
      const shouldRemoveFilter = valueShouldBeRemoved(filteringValue);

      if (shouldRemoveFilter) {
        return {
          ...state,
          navigationPropertyFilters: {
            ...removeKey(filteringProperty, state.navigationPropertyFilters),
          },
        };
      }

      return {
        ...state,
        navigationPropertyFilters: {
          ...state.navigationPropertyFilters,
          [filteringProperty]: {
            value: filteringValue,
            navigationProperty,
            generatedExpression,
            navigationPropertyIsNested,
          },
        },
      };
    }
    case UPDATE_FUNCTION_FILTER: {
      const { filteringProperty, filterQueryString, values } = action;
      const shouldRemoveFilter = filterQueryString === null;

      if (shouldRemoveFilter) {
        return {
          ...state,
          functionFilters: state.functionFilters.filter(
            (f) => f.filteringProperty !== filteringProperty
          ),
        };
      }

      const onlyUpdateExisting = state.functionFilters.some(
        (f) => f.filteringProperty === filteringProperty
      );

      if (onlyUpdateExisting) {
        return {
          ...state,
          functionFilters: state.functionFilters.map((f) =>
            f.filteringProperty === filteringProperty
              ? { ...f, queryString: filterQueryString, values }
              : f
          ),
        };
      }

      // Insert new one
      return {
        ...state,
        functionFilters: [
          ...state.functionFilters,
          {
            filteringProperty,
            queryString: filterQueryString,
            values,
          },
        ],
      };
    }
    case UPDATE_MULTIPLE_FUNCTION_FILTERS: {
      const { filtersToUpdate } = action;

      filtersToUpdate.forEach((filter: any) => {
        const { filteringProperty, filterQueryString, values } = filter;

        if (filterQueryString) {
          const updateExisting = state.functionFilters.some(
            (f) => f.filteringProperty === filteringProperty
          );
          if (updateExisting)
            return {
              ...state,
              functionFilters: state.functionFilters.map((f) =>
                f.filteringProperty === filteringProperty
                  ? { ...f, queryString: filterQueryString, values }
                  : f
              ),
            };
          else
            return {
              ...state,
              functionFilters: [
                ...state.functionFilters,
                {
                  filteringProperty,
                  queryString: filterQueryString,
                  values,
                },
              ],
            };
        }
        // Remove
        else
          return {
            ...state,
            functionFilters: state.functionFilters.filter(
              (f) => f.filteringProperty !== filteringProperty
            ),
          };
      });

      return state;
    }
    case RESET_ALL_FILTERS: {
      return {
        ...state,
        ...containerInitialState,
      };
    }
    case RESET_POPERTY_FILTERS: {
      const filteringProperties: any[] = action.filteringProperties;

      let modifiedPropertyFiltetrs = cloneDeep(state.propertyFilters);

      filteringProperties.forEach((filter) => {
        if (state.propertyFilters[filter])
          modifiedPropertyFiltetrs = removeKey(
            filter,
            modifiedPropertyFiltetrs
          );
      });
      return {
        ...state,
        propertyFilters: modifiedPropertyFiltetrs,
      };
    }
    case OVERRIDE_STATE: {
      const { filterState } = action;

      return {
        propertyFilters:
          filterState?.propertyFilters ?? containerInitialState.propertyFilters,
        navigationPropertyFilters:
          filterState?.navigationPropertyFilters ??
          containerInitialState.navigationPropertyFilters,
        functionFilters:
          filterState?.functionFilters ?? containerInitialState.functionFilters,
      };
    }
    default:
      return state;
  }
};

// All helper methods process only data from filter instances, created with this filter reducer
export class FilterHelperMethods {
  /**
   * Method will use propertyFilters, navigationPropertyFilters and functionFilters(objects and array from filter instance) to generate kendo filter object that GenericDataGrid can use when calling getMethod, in order to filter data
   * @param {Object} filters propertyFilters map - part of filter instance state
   * @param {Object[]} functionFilters functionFilters array - part of filter instance state
   * @param {Object} navigationPropertyFilters navigationPropertyFilters map - part of filter instance state
   */
  static getFilterStrings = (
    filters: any,
    functionFilters = null,
    navigationPropertyFilters = null
  ) => {
    const propertyFilters: any = {
      logic: "and",
      filters: [],
    };

    const filterQueryStrings: any = {
      propertyFiltersQueryString: "",
      functionFiltersQueryString:
        FilterHelperMethods.getFunctionFiltersQueryString(functionFilters),
      navigationPropertyFilterQueryString:
        FilterHelperMethods.getNavigationPropFiltersQueryString(
          navigationPropertyFilters
        ),
    };

    Object.keys(filters).forEach((k) => {
      const filterValue = filters[k].value;
      const operator = filters[k].operator ?? "eq";
      const filterType = filters[k].type;
      const logic = filters[k].logic ?? "or";

      if (
        !Array.isArray(filterValue) &&
        (filterValue instanceof Date || filterType === "datetime")
      ) {
        const value =
          typeof filterValue === "string" ? new Date(filterValue) : filterValue; // Need this check because when parsing stringified object from local storage then Datetime object becomes string so we need to convert it back to Date object
        propertyFilters.filters.push({
          field: k,
          value,
          operator,
          key: k,
        });
      } else if (
        typeof filterValue === "string" ||
        typeof filterValue === "number" ||
        typeof filterValue === "boolean" ||
        filterValue === null
      ) {
        if (k.includes(";")) {
          // Multi-filter (filtering more than one fields with one filter by using field name like 'Field1;Field2;Field3')
          const multiFilter: any = {
            key: k,
            filters: [],
            logic,
          };
          propertyFilters.filters.push(multiFilter);

          k.split(";").forEach((field) => {
            multiFilter.filters.push({
              field,
              value: filterValue,
              operator,
            });
          });
        } else
          propertyFilters.filters.push({
            field: k,
            value: filterValue,
            operator,
            key: k,
          });
      } else if (Array.isArray(filterValue) && filterValue.length > 0) {
        propertyFilters.filters.push({
          key: k,
          filters: [],
          logic,
        });
        filterValue.forEach((v, ind) => {
          const value = v.value ? v.value : v;
          const isDateTime = filterType === "datetime";
          const multipleOperators = Array.isArray(operator);
          const filterOperator = multipleOperators ? operator[ind] : operator;

          const proccessedValue =
            isDateTime && typeof value === "string" ? new Date(value) : value; // Need this check because when parsing stringified object from local storage then Datetime object becomes string so we need to convert it back to Date object
          const index = propertyFilters.filters.findIndex(
            (f: any) => f.key === k
          );

          if (index >= 0) {
            let filterToAssign = {
              field: k,
              value: proccessedValue,
              operator: filterOperator,
            };
            if (v.value) filterToAssign = { ...filterToAssign, ...v };
            propertyFilters.filters[index].filters.push(filterToAssign);
          }
        });
      }
    });

    filterQueryStrings.propertyFiltersQueryString =
      serializeFilter(propertyFilters);

    return filterQueryStrings;
  };

  /**
   * Method will use navigationPropertyFilters map(from filter instance) to generate query string(with lambdas (and lambda custom expressions)). Should be used in conjunction with upper method and then both results should be passed, as args, to toODataStr method to get complete query string for grid method request
   * @param {Object} navigationPropertyFilters
   */
  static getNavigationPropFiltersQueryString = (
    navigationPropertyFilters: any
  ) => {
    if (!navigationPropertyFilters || navigationPropertyFilters.length === {})
      return null;
    let groupedFilters: any = [];
    let query = "";

    Object.keys(navigationPropertyFilters).forEach((f) => {
      const filter = navigationPropertyFilters[f];

      if (
        groupedFilters.some(
          (gf: any) => gf.navigationProperty === filter.navigationProperty
        )
      ) {
        groupedFilters = groupedFilters.map((gf: any) => {
          if (gf.navigationProperty === filter.navigationProperty) {
            return {
              navigationProperty: filter.navigationProperty,
              values: filter.generatedExpression
                ? gf.values
                : [...gf.values, { field: f, value: filter.value }],
              generatedExpressions: filter.generatedExpression
                ? [...gf.generatedExpressions, filter.generatedExpression]
                : gf.generatedExpressions,
              navigationPropertyIsNested: filter.navigationPropertyIsNested,
            };
          }
          return gf;
        });
      } else {
        groupedFilters.push({
          navigationProperty: filter.navigationProperty,
          values: filter.generatedExpression
            ? []
            : [{ field: f, value: filter.value }],
          generatedExpressions: filter.generatedExpression
            ? [filter.generatedExpression]
            : [],
          navigationPropertyIsNested: filter.navigationPropertyIsNested,
        });
      }
    });

    if (groupedFilters.length > 0) {
      groupedFilters.forEach((f: any, ind: number) => {
        const multipleLevelNavigationProperty =
          f.navigationProperty.indexOf("/") >= 0 &&
          f.navigationPropertyIsNested;

        const splittedName = f.navigationProperty.split("/");
        const navPropName = multipleLevelNavigationProperty
          ? splittedName.slice(0, splittedName.length - 1).join("/")
          : f.navigationProperty;
        const subNavPropName = multipleLevelNavigationProperty
          ? splittedName[splittedName.length - 1]
          : null;

        // Here we can see one limitation of this helper method => we are forced to use 's' as input parameter in lambda expression when populating filters
        // We can add mechanism to support any other letter but since the scope of this method is not big(only few of us devs are using this), then we should not add additional complexity just for that purpose
        if (multipleLevelNavigationProperty)
          query += `${navPropName}/any(x: x/${subNavPropName}/any(s: {expressionPart}))`;
        else query += `${navPropName}/any(s: {expressionPart})`;
        let expressionPart = "";

        if (f.generatedExpressions && f.generatedExpressions.length > 0) {
          f.generatedExpressions.forEach((ce: any, i: number) => {
            if (i === f.generatedExpressions.length - 1)
              expressionPart += ` ${ce}`;
            else expressionPart += ` ${ce} and`;
          });
        }

        if (f.values && f.values.length > 0) {
          if (expressionPart) expressionPart += " and";
          f.values.forEach((v: any, index: number) => {
            if (Array.isArray(v.value) && v.value.length > 0) {
              expressionPart += ` (`;
              v.value.forEach((vv: any, i: number) => {
                let value = vv.value ? vv.value : vv;
                if (!isBoolean(value)) value = `'${value}'`;
                if (i === v.value.length - 1)
                  expressionPart += ` s/${v.field} eq ${value})`;
                else expressionPart += ` s/${v.field} eq ${value} or`;
              });
            } else {
              let value = v.value;
              if (!isBoolean(value)) value = `'${value}'`;
              expressionPart += ` (s/${v.field} eq ${value})`;
            }
            if (index < f.values.length - 1) expressionPart += " and";
          });
        }

        query = query.replace("{expressionPart}", expressionPart);
        if (ind < groupedFilters.length - 1) query += " and ";
      });
    }
    return query;
  };

  /**
   * Concats function filter strings into one string
   * @param {Object[]} functionFilters function filters array - part of filter instance
   */
  static getFunctionFiltersQueryString = (functionFilters: any) => {
    let queryString = "";
    if (Array.isArray(functionFilters) && functionFilters.length > 0) {
      functionFilters.forEach((ff, ind) => {
        if (ind === 0) queryString += `${ff.queryString}`;
        else queryString += ` and ${ff.queryString}`;
      });
    }
    return queryString;
  };

  static generateODataFilterString = (filters: any) => {
    const {
      propertyFiltersQueryString,
      navigationPropertyFilterQueryString,
      functionFiltersQueryString,
    } = FilterHelperMethods.getFilterStrings(
      filters.propertyFilters,
      filters.functionFilters,
      filters.navigationPropertyFilters
    );

    const odataFilterQuery = concatFilterQuerySubstrings(
      propertyFiltersQueryString,
      navigationPropertyFilterQueryString,
      functionFiltersQueryString
    );

    return odataFilterQuery.replace(
      /('[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}')/gi,
      (x) => x.substring(1, x.length - 1)
    );
  };
}
