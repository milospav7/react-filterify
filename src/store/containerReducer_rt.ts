import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { valueShouldBeRemoved } from "./helpers";
import { Container } from "./types";

const containerInitialState: Container = {
  propertyFilters: {},
  navigationPropertyFilters: {},
  functionFilters: [],
};

interface IPayload {
  filteringProperty: string;
  filteringValue: any;
}

interface IPropertyFilterUpdate extends IPayload {
  operator: string;
  logic: string;
  allowNullValue: boolean;
}

interface INavigationPropertyFilterUpdate extends IPayload {
  navigationProperty: string;
  generatedExpression: string;
}

interface IFunctionFilterUpdate {
  filteringProperty: string;
  filterQueryString: string;
  values: any;
}

const editForms = createSlice({
  name: "filterify_container",
  initialState: containerInitialState,
  reducers: {
    updatePropertyFilter(state, action: PayloadAction<IPropertyFilterUpdate>) {
      const {
        filteringProperty,
        filteringValue,
        operator,
        logic,
        allowNullValue,
      } = action.payload;
      const shouldRemoveFilter =
        !allowNullValue && valueShouldBeRemoved(filteringValue);

      if (shouldRemoveFilter) {
        delete state.propertyFilters[filteringProperty];
      } else {
        state.propertyFilters[filteringProperty] = {
          value: filteringValue,
          operator,
          logic,
        };
      }
    },
    updateNavigationPropertyFilter(
      state,
      action: PayloadAction<INavigationPropertyFilterUpdate>
    ) {
      const {
        navigationProperty,
        filteringProperty,
        filteringValue,
        generatedExpression,
      } = action.payload;
      const shouldRemoveFilter = valueShouldBeRemoved(filteringValue);

      if (shouldRemoveFilter) {
        delete state.navigationPropertyFilters[filteringProperty];
      } else {
        state.navigationPropertyFilters[filteringProperty] = {
          value: filteringValue,
          navigationProperty,
          generatedExpression,
        };
      }
    },
    updateFunctionFilter(state, action: PayloadAction<IFunctionFilterUpdate>) {
      const { filteringProperty, filterQueryString, values } = action.payload;
      const shouldRemoveFilter = filterQueryString === null;
      const onlyUpdateExisting = state.functionFilters.some(
        (f) => f.filteringProperty === filteringProperty
      );

      if (shouldRemoveFilter) {
        state.functionFilters = state.functionFilters.filter(
          (f) => f.filteringProperty !== filteringProperty
        );
      } else if (onlyUpdateExisting) {
        state.functionFilters = state.functionFilters.map((f) =>
          f.filteringProperty === filteringProperty
            ? { ...f, queryString: filterQueryString, values }
            : f
        );
      } else {
        // Insert new one
        state.functionFilters.push({
          filteringProperty,
          queryString: filterQueryString,
          values,
        });
      }
    },
    resetAllFilters(state) {
      return { ...state, ...containerInitialState };
    },
  },
});

// Extract the action creators object and the reducer
const { actions, reducer } = editForms;

// Extract and export each action creator by name
export const { updatePropertyFilter } = actions;

// Export the reducer, either as a default or named export
