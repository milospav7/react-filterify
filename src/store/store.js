import { createStore, combineReducers } from "redux";
import { composeWithDevTools } from "redux-devtools-extension";
import { configureFilterfyReducer } from "./orchestratorReducer";

export const CONTAINER_IDS = {
  C1: "C1",
  C2_Test: "C2_Test",
};

const composedEnhancers = composeWithDevTools();
const filterifyFilters = configureFilterfyReducer([
  {
    id: CONTAINER_IDS.C1,
    saveToLocalStorage: true,
    styleSchema: {
      highlightWhenInUse: true,
      labelColor: "red",
      labelFontSize: "13px",
    },
  },
  {
    id: CONTAINER_IDS.C2_Test,
    saveToLocalStorage: false,
  },
]);

const reducers = combineReducers({
  filterifyFilters,
});

const store = createStore(reducers, composedEnhancers);

export default store;
