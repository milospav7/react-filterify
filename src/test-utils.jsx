// test-utils.jsx
import { render as rtlRender } from "@testing-library/react";
import { Provider } from "react-redux";
import appStore from "./store/store";

function render(
  ui,
  { preloadedState, store = appStore, ...renderOptions } = {}
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>;
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

// re-export everything
export * from "@testing-library/react";
// override render method
export { render };
