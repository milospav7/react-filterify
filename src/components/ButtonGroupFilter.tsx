import React, { useCallback, useMemo } from "react";
import { useDispatch } from "react-redux";
import { Button, ButtonGroup } from "reactstrap";
import { updatePropertyFilter } from "../store/actionCreators";
import { useFilterifyFilter } from "./hooks";

interface IProps {
  containerId: string;
  filteringProperty: string;
  options: Array<any>;
  isMulti?: boolean;
  wrapperClassName?: string;
}

const ButtonGroupFilter: React.FC<IProps> = ({
  containerId,
  filteringProperty,
  options,
  isMulti = false,
  wrapperClassName,
}) => {
  const { propertyFilters } = useFilterifyFilter(containerId);
  const currentFilterValue = propertyFilters[filteringProperty]?.value;
  const filterOperaetor = propertyFilters[filteringProperty]?.operator;
  const dispatcher = useDispatch();

  const updateInMultiValueMode = useCallback(
    (value, operator, logic) => {
      if (currentFilterValue) {
        let newValue = currentFilterValue.some((v: any) => v === value)
          ? currentFilterValue.filter((v: any) => v !== value)
          : [...currentFilterValue, value];

        dispatcher(
          updatePropertyFilter(
            containerId,
            filteringProperty,
            newValue,
            operator,
            logic
          )
        );
      } else
        dispatcher(
          updatePropertyFilter(
            containerId,
            filteringProperty,
            [value],
            operator,
            logic
          )
        );
    },
    [containerId, dispatcher, filteringProperty, currentFilterValue]
  );

  const updateInSingleValueMode = useCallback(
    (value, operator, logic) => {
      if (currentFilterValue !== undefined && operator === filterOperaetor)
        dispatcher(updatePropertyFilter(containerId, filteringProperty, null));
      else
        dispatcher(
          updatePropertyFilter(
            containerId,
            filteringProperty,
            value,
            operator,
            logic
          )
        );
    },
    [
      containerId,
      currentFilterValue,
      dispatcher,
      filteringProperty,
      filterOperaetor,
    ]
  );

  const setPropertyFilter = useCallback(
    (value, operator = "eq", logic = "or") => {
      if (isMulti) {
        updateInMultiValueMode(value, operator, logic);
      } else {
        updateInSingleValueMode(value, operator, logic);
      }
    },
    [isMulti, updateInMultiValueMode, updateInSingleValueMode]
  );

  const isOptionSelected = useCallback(
    (option, optOperator = "eq") => {
      if (isMulti)
        return !!currentFilterValue?.some(
          (v: any) => v === option && optOperator === filterOperaetor
        );
      return currentFilterValue === option && optOperator === filterOperaetor;
    },
    [currentFilterValue, isMulti, filterOperaetor]
  );

  const memoizedFilter = useMemo(
    () => (
      <div className={wrapperClassName}>
        <ButtonGroup key={`bgf-${filteringProperty}`} size="sm">
          {options.map((opt, ind) => (
            <Button
              key={`${ind}-${filteringProperty}`}
              className="no-higlight-focus"
              color="light"
              onClick={() => {
                setPropertyFilter(opt);
              }}
              active={isOptionSelected(opt)}
            >
              {opt}
            </Button>
          ))}
        </ButtonGroup>
      </div>
    ),
    [
      wrapperClassName,
      filteringProperty,
      options,
      isOptionSelected,
      setPropertyFilter,
    ]
  );

  return memoizedFilter;
};

export default ButtonGroupFilter;
