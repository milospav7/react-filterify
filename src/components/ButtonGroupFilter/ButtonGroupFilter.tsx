/* eslint-disable react-hooks/exhaustive-deps */
import React, { useCallback, useMemo } from "react";
import { useDispatch } from "react-redux";
import { Button, ButtonGroup } from "reactstrap";
import { BaseFilterProps } from "../../store/interfaces";
import FilterDecorator from "../shared/FilterDecorator";
import {
  useContainerStyleSchema,
  useFilterActions,
  useFilterState,
} from "../../hooks";

interface IProps extends BaseFilterProps {
  options: Array<any>;
  isMulti?: boolean;
  logic?: "or" | "and";
}

const ButtonGroupFilter: React.FC<IProps> = ({
  containerId,
  filteringProperty,
  navigationProperty,
  options,
  isMulti = false,
  className,
  withAssociatedLabel,
  labelClassName,
  label,
  style,
  logic = "or",
  size = "sm",
}) => {
  const dispatcher = useDispatch();

  const { updateFilter } = useFilterActions(
    containerId,
    filteringProperty,
    navigationProperty
  );
  const { filterValue, filterOperator } = useFilterState(
    containerId,
    filteringProperty,
    navigationProperty
  );
  const { styles } = useContainerStyleSchema(containerId);

  const updateInMultitMode = useCallback(
    (value) => {
      let modifiedFilter;
      const previousFilter = filterValue ?? [];
      const allOptionsDeselected =
        previousFilter.length === 1 && previousFilter[0] === value;

      if (allOptionsDeselected) modifiedFilter = null;
      else
        modifiedFilter = previousFilter.some((v: any) => v === value)
          ? previousFilter.filter((v: any) => v !== value)
          : [...previousFilter, value];

      updateFilter(modifiedFilter, { operator: "eq", logic });
    },
    [containerId, dispatcher, filteringProperty, filterValue]
  );

  const updateInSingleMode = useCallback(
    (value) => {
      const modifiedFilter = filterValue === value ? null : value;
      updateFilter(modifiedFilter, { operator: "eq", logic });
    },
    [containerId, filterValue, dispatcher, filteringProperty, filterOperator]
  );

  const updateTargetFilter = useCallback(
    (value) => {
      if (isMulti) {
        updateInMultitMode(value);
      } else {
        updateInSingleMode(value);
      }
    },
    [isMulti, updateInMultitMode, updateInSingleMode]
  );

  const isOptionSelected = useCallback(
    (option) => {
      if (isMulti) return filterValue?.some((v: any) => v === option);
      return filterValue === option;
    },
    [filterValue, isMulti]
  );

  const memoizedFilter = useMemo(
    () => (
      <FilterDecorator
        displayLabel={withAssociatedLabel}
        className={className}
        labelClassName={labelClassName}
        label={label}
        style={style}
        labelStyle={styles.label}
      >
        <ButtonGroup key={`bgf-${filteringProperty}`} size={size}>
          {options.map((opt, ind) => (
            <Button
              key={`${ind}-${filteringProperty}`}
              className="no-higlight-focus"
              color="light"
              onClick={() => {
                updateTargetFilter(opt);
              }}
              active={isOptionSelected(opt)}
            >
              {opt}
            </Button>
          ))}
        </ButtonGroup>
      </FilterDecorator>
    ),
    [
      className,
      filteringProperty,
      options,
      isOptionSelected,
      updateTargetFilter,
    ]
  );

  return memoizedFilter;
};

export default ButtonGroupFilter;
