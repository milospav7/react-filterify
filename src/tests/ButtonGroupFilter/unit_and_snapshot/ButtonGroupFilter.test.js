import { screen } from "@testing-library/react";
import ButtonGroupFilter from "../../../components/ButtonGroupFilter/ButtonGroupFilter";
import { CONTAINER_IDS } from "../../../store/store";
import { render, rtrCreate } from "../../../test-utils";

/** UNIT TESTING */
test("Button group filter with label", () => {
  render(
    <ButtonGroupFilter
      containerId={CONTAINER_IDS.C2_Test}
      withAssociatedLabel
      label="Region"
      isMulti
      filteringProperty="Region"
      options={["Šumadija", "Banat", "Srem"]}
      className="text-start mb-3"
    />
  );
  expect(screen.getByText(/region/i)).toBeInTheDocument();
  expect(screen.getByRole("button", { name: /šumadija/i })).toBeInTheDocument();
  expect(screen.getByRole("button", { name: /banat/i })).toBeInTheDocument();
  expect(screen.getByRole("button", { name: /srem/i })).toBeInTheDocument();
});

test("Button group filter without label", () => {
  render(
    <ButtonGroupFilter
      containerId={CONTAINER_IDS.C2_Test}
      isMulti
      filteringProperty="Region"
      options={["Šumadija", "Banat", "Srem"]}
      className="text-start mb-3"
    />
  );
  expect(screen.queryByText(/region/i)).not.toBeInTheDocument(); // ** getByText will throw error if nothing found
  expect(screen.getByRole("button", { name: /šumadija/i })).toBeInTheDocument();
  expect(screen.getByRole("button", { name: /banat/i })).toBeInTheDocument();
  expect(screen.getByRole("button", { name: /srem/i })).toBeInTheDocument();
});

// To throw error without container id

/** SNAPSHOT TESTING */
it("Render ButtonGroupFilter as expected", () => {
  const filter = rtrCreate(
    <ButtonGroupFilter
      containerId={CONTAINER_IDS.C2_Test}
      isMulti
      filteringProperty="Region"
      options={["Šumadija", "Banat", "Srem"]}
      className="text-start mb-3"
    />
  ).toJSON();

  expect(filter).toMatchSnapshot();
});
