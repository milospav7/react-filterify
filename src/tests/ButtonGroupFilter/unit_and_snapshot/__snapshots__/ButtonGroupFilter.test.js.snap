// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render ButtonGroupFilter as expected 1`] = `
<div
  className="text-start mb-3"
>
  <div
    className="btn-group-sm btn-group"
    role="group"
  >
    <button
      aria-label={null}
      className="no-higlight-focus btn btn-light"
      onClick={[Function]}
      type="button"
    >
      Å umadija
    </button>
    <button
      aria-label={null}
      className="no-higlight-focus btn btn-light"
      onClick={[Function]}
      type="button"
    >
      Banat
    </button>
    <button
      aria-label={null}
      className="no-higlight-focus btn btn-light"
      onClick={[Function]}
      type="button"
    >
      Srem
    </button>
  </div>
</div>
`;
